### ISSUE TITLE
[Feature request] Send raw json event

### ISSUE THREAD

**Issue by opk12** [8go/matrix-commander#60]
Edits, replies and reactions are not supported. Other message relations are in the works, with varying support depending on the client (threads, generalized relations). Rather than opening an issue for each feature, and wait for this codebase to be updated, it's quicker for me to template-ize a real-world example into a script that emits json, to be piped into matrix-commander.

**Comment by 8go** [comment]
Hi @opk12 Thanks for your idea and suggestion.

Nice idea. 

I am considering implementing it. 

Can you provide some real JSON objects that I can use for testing? Maybe 1 or 2 simple ones, and 1 or 2 more complex ones. Please assist. 

How do you envision the CLI? 
- JSON file?  `matrix-commander.py --json somejsonfile.json ...`
- Piping it in? some-generator | matrix-commander.py --json ...`

**Comment by 8go** [comment]
I did the piping part for images this week, so both file and pipe should be fine for future `json` data.  
it would be 
`matrix-commander.py --json somejsonfile.json ...`
`some-generator | matrix-commander.py --json - ...`

Any sample json files?

**Comment by opk12** [comment]
What about `--raw-event` or more explicit `--send-event`? `--json` may be reserved to switch the output format for `--listen`, and could read as auth/config data to someone who has no experience of this command.

I researched message events. In general, does the feature allow to send state events, for example changing the room topic? I do not understand if matrix-nio makes it possible.

```
#!/bin/bash

# Pasted from the MSC.
# %s will be replaced by printf.
JSON_EDIT_MSC2676='{ "type": "m.room.message", "content": { "body": "%s", "msgtype": "m.text", "m.new_content": { "body": "%s", "msgtype": "m.text" }, "m.relates_to": { "rel_type": "m.replace", "event_id": "%s" } } }'
JSON_REACT_MSC2677='{ "type": "m.reaction", "content": { "m.relates_to": { "rel_type": "m.annotation", "event_id": "%s", "key": "%s" } } }'
JSON_REPLY_AS_THREAD_MSC3440='{ "type": "m.room.message", "content": { "body": "%s", "msgtype": "m.text", "m.relates_to": { "rel_type": "m.thread", "event_id": "%s" } } }'

TARGET_EVENT='$some event id'

printf "$JSON_EDIT_MSC2676" "Fallback body $(date +%H:%M:%S)" "Non-fallback body $(date +%H:%M:%S)" "$TARGET_EVENT" \
| matrix-commander ... --raw-event -

printf "$JSON_REACT_MSC2677" "$TARGET_EVENT" "👍" \
| matrix-commander ... --raw-event -

printf "$JSON_REPLY_AS_THREAD_MSC3440" "Thread reply $(date +%H:%M:%S)"  "$TARGET_EVENT" \
| matrix-commander ... --raw-event -

```

I tested with

```
python3 -m asyncio
    import nio
    client = AsyncClient(...)
    # .login(), .load_store(), .sync()
    content_json = json.loads(piped_input)
    message_type = content_json['type']
    content_inner = content_json['content']
    await client.room_send(room_id=room_id, message_type=message_type, content=content_inner)
```

Bash string syntax ([reference here](https://www.gnu.org/software/bash/manual/html_node/Quoting.html#Quoting))
- `'string'` is literal characters without special behavior for inner `" $ \` and newline. Analogous to python `r"""string"""`.
- `"string"` outputs one string and expands any variables. Analogous to python `f"""string"""`.

**Comment by 8go** [comment]
:clap:  Thank you! It is appreciated.

Your comment about the naming is valid. `--json` is not so meaningful. Based on your feedback I have a couple of additional alternatives: `--raw` and `--send-raw` and `--event`. Comments?  Preferences?

In context: `--image` (not `--send-image`), `--audio` (not `--send-audio`), etc: it would make it inconsistent if we start using `--send-xxx` now. Maybe better: `--raw` or `--event` ?

I sort of lean more towards `--event`? Your thoughts?

**Comment by opk12** [comment]
I agree. `--event` is focused on the "what" rather than the format, and fits with image, audio, message, location, ...

**Comment by 8go** [comment]
Just a quick response regarding point (2). 

I think we should make this a new separate issue. I created one for it. See Issue #72.

Please go to Issue #72 and give your opinion/suggestions there. Thank you.

**Comment by 8go** [comment]
Quick comment on (1). As you pointed out, definitely a bug: `read()` is being called twice. Will fix it tomorrow.

**Comment by 8go** [comment]
Bug, as described in (1) is fixed in commit [05512ec](https://github.com/8go/matrix-commander/commit/05512ec10cc88424cc8becfb98c13b990d75196f)

@opk12 Have a look at the new version if you have time.  :heart:

**Comment by 8go** [comment]
@opk12 I have a question for you.

Which URL would you recommend for reading to the user who wants to use events and create  JSON event objects? Best URL for Matrix MSC descriptions for events?

I have seen this and a few others: https://spec.matrix.org/unstable/proposals/

Are there better ones to put into the documentation/help ?

**Comment by opk12** [comment]
> Please go to Issue https://github.com/8go/matrix-commander/issues/72 and give your opinion/suggestions there.

Thank you! I left some thoughts there.

> Also added commit https://github.com/8go/matrix-commander/commit/66a5a62b5b1dea3f104bdc9051342dae40917c8f that added your event templates in a nicer, hopefully more re-usable way. So, that other people can reuse better and more.

This informs the user and allows to share the code. Cool idea!

> I have seen this and a few others: https://spec.matrix.org/unstable/proposals/
> Are there better ones to put into the documentation/help ?

I do not know a better link. There is the GitHub MSC repo, but it is already reachable through your link, might not have all the MSCs and has the superseded proposals.

**Comment by 8go** [comment]
Thanks again @opk12 for the feedback and all the very useful and helpful input :clap:  :clap:  :clap: . Excellent. You definitely added a lot of value. 

Can the "event" issue be closed? Or are there other event related thoughts that should be added to the code?

Beyond events, generally speaking: It seems you have a keen eye to spot weak code or non-python-esque code. Feel free to comment or make PR to make the code better.

**Comment by 8go** [comment]
> And thank you for adding me to the contributors!

No need to thank me. you EARNED that.
